File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/operation/meta/value/factory/StringValueFactory.java
package ru.citeck.ecos.records3.record.operation.meta.value.factory;

import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.records3.RecordRef;
import ru.citeck.ecos.records3.record.operation.meta.value.AttValue;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

@Slf4j
public class StringValueFactory implements AttValueFactory<String> {

    private final Map<String, Function<String, ?>> converters = new ConcurrentHashMap<>();

    public StringValueFactory() {
        converters.put("ref", RecordRef::valueOf);
    }

    @Override
    public AttValue getValue(String value) {

        return new AttValue() {
            @Override
            public String getString() {
                return value;
            }

            @Override
            public Boolean getBool() {
                return Boolean.TRUE.toString().equals(value);
            }

            @Override
            public Double getDouble() {
                return Double.parseDouble(value);
            }

            @Override
            public Object getAs(@NotNull String type) {
                Function<String, ?> converter = converters.get(type);
                return converter != null ? converter.apply(value) : null;
            }
        };
    }

    public void addConverter(String type, Function<String, ?> converter) {
        Function<String, ?> currentConverter = this.converters.get(type);
        if (currentConverter != null) {
            log.warn("Converter with type " + type + " (" + currentConverter + ") will be replaced by " + converter);
        }
        this.converters.put(type, converter);
    }

    @Override
    public List<Class<? extends String>> getValueTypes() {
        return Collections.singletonList(String.class);
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/atts/value/factory/StringValueFactory.kt
package ru.citeck.ecos.records3.record.atts.value.factory

import io.github.oshai.kotlinlogging.KotlinLogging
import ru.citeck.ecos.commons.data.MLText
import ru.citeck.ecos.records3.record.atts.value.AttValue
import ru.citeck.ecos.webapp.api.entity.EntityRef
import java.lang.NumberFormatException
import java.util.concurrent.ConcurrentHashMap

@Suppress("unused")
class StringValueFactory : AttValueFactory<String> {

    companion object {
        private val log = KotlinLogging.logger {}
    }

    private val converters = ConcurrentHashMap<String, (String) -> Any?>()

    init {
        converters["ref"] = { EntityRef.valueOf(it) }
        converters["mltext"] = { MLText(it) }
    }

    override fun getValue(value: String): AttValue {

        return object : AttValue {

            override fun getId(): Any {
                return value
            }

            override fun asText(): String {
                return value
            }

            override fun asDouble(): Double? {
                if (value.isEmpty()) {
                    return null
                }
                return try {
                    value.toDouble()
                } catch (e: NumberFormatException) {
                    null
                }
            }

            override fun getAs(type: String): Any? {
                return converters[type]?.invoke(value)
            }

            override fun asRaw(): Any {
                return value
            }

            override fun asBin(): Any {
                return value
            }
        }
    }

    fun addConverter(type: String, converter: (String) -> Any?) {
        val currentConverter = converters[type]
        if (currentConverter != null) {
            log.warn { "Converter with type $type ($currentConverter) will be replaced by $converter" }
        }
        converters[type] = converter
    }

    override fun getValueTypes() = listOf(String::class.java)
}
